python/ray/serve/master.py

```diff
@@ -1,334 +1,528 @@
-        ).remote(host, port)
-
-    async def get_http_proxy_config(self):
-        route_table = self.route_table.list_service(
-            include_methods=True, include_headless=False)
-        return route_table, self.get_router()
+        try:
+            self.http_proxy = ray.util.get_actor(SERVE_PROXY_NAME)
+        except ValueError:
+            logger.info(
+                "Starting HTTP proxy with name '{}'".format(SERVE_PROXY_NAME))
+            self.http_proxy = async_retryable(HTTPProxyActor).options(
+                detached=True,
+                name=SERVE_PROXY_NAME,
+                max_concurrency=ASYNC_CONCURRENCY,
+                max_reconstructions=ray.ray_constants.INFINITE_RECONSTRUCTION,
+            ).remote(host, port)
 
     def get_http_proxy(self):
-        assert self.http_proxy is not None, "HTTP proxy not started yet."
+        """Returns a handle to the HTTP proxy managed by this actor."""
         return [self.http_proxy]
 
-    def start_metric_monitor(self, gc_window_seconds):
-        assert self.metric_monitor is None, "Metric monitor already started."
-        self.metric_monitor = MetricMonitor.remote(gc_window_seconds)
-        # TODO(edoakes): this should be an actor method, not a separate task.
-        start_metric_monitor_loop.remote(self.metric_monitor)
-        self.metric_monitor.add_target.remote(self.router)
+    def get_http_proxy_config(self):
+        """Called by the HTTP proxy on startup to fetch required state."""
+        return self.routes, self.get_router()
+
+    def _get_or_start_metric_monitor(self, gc_window_s):
+        """Get the metric monitor belonging to this serve cluster.
+
+        If the metric monitor does not already exist, it will be started.
+        """
+        try:
+            self.metric_monitor = ray.util.get_actor(SERVE_METRIC_MONITOR_NAME)
+        except ValueError:
+            logger.info("Starting metric monitor with name '{}'".format(
+                SERVE_METRIC_MONITOR_NAME))
+            self.metric_monitor = MetricMonitor.options(
+                detached=True,
+                name=SERVE_METRIC_MONITOR_NAME).remote(gc_window_s)
+            # TODO(edoakes): move these into the constructor.
+            start_metric_monitor_loop.remote(self.metric_monitor)
+            self.metric_monitor.add_target.remote(self.router)
 
     def get_metric_monitor(self):
-        assert self.metric_monitor is not None, (
-            "Metric monitor not started yet.")
+        """Returns a handle to the metric monitor managed by this actor."""
         return [self.metric_monitor]
 
+    def _checkpoint(self):
+        """Checkpoint internal state and write it to the KV store."""
+        logger.debug("Writing checkpoint")
+        start = time.time()
+        checkpoint = pickle.dumps(
+            (self.routes, self.backends, self.traffic_policies, self.replicas,
+             self.replicas_to_start, self.replicas_to_stop))
```

Comment:
Hmm I think this might be nice if we could use it for both writing and reading the checkpoint data, but I can't think of any straightforward way to do that. This also has the benefit of being more clear about what data is being written/loaded in `self._checkpoint` and `self._recover_from_checkpoint` without needing to go look at another variable definition